### PR TITLE
added some tests for dashboard.py

### LINKED ISSUE #101: Add unit tests to streamlit
# Description
Add more unit tests to dashboard.py to ensure functionality of the script.

# Outcome
- Increase pytest coverage of streamlit dashboard folder
- Add unit tests to dashboard.py to ensure functionality and consider edge cases

# User Story
As a user, when I subscribe/unsubscribe to products, I want to be able to track the prices without any issues or errors.

### PR DESCRIPTION
**Pull request by gem09lo**
- Added some tests for streamlit dashboard.py

This partiall closes #101

### CODE REVIEW

**Review by indiah444 — CHANGES_REQUESTED** [review]
Please delete the commented out function if it's not being used.

File: streamlit_dashboard/test_database_connection.py [hunk 1]
```diff
@@ -0,0 +1,465 @@
...
+            mock_cursor = MagicMock()
+            mock_get_conn.return_value = mock_connection
+            mock_get_cursor.return_value = mock_cursor
+            mock_cursor.execute.side_effect = psycopg2.Error("Database error")
+
+            result = get_product_subscription(1)
+            assert result == None
+
+
+# 95% - %75
```
Comment (indiah444) [inline streamlit_dashboard/test_database_connection.py:465]:
```suggestion
```

File: streamlit_dashboard/test_dashboard.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+import altair as alt
+from bs4 import BeautifulSoup
+from dashboard import (
+    get_html_from_url,
+    get_website_from_url,
+    display_charts, login, track_product, login_clicked
+)
+
+
+@pytest.fixture
```
Comment (indiah444) [inline streamlit_dashboard/test_dashboard.py:17]: Love the use of fixtures!

File: streamlit_dashboard/test_dashboard.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    mock_st.selectbox.assert_called_once_with(
+        "Select Time Range", ["Last 3 Days",
+                              "Last 24 Hours", "Last 30 Minutes"]
+    )
+
+
+@patch('dashboard.st')
+@patch('dashboard.execute_database_select_query_fetchall')
+def test_empty_filtered_df(mock_db_function, mock_st):
+    """Test when the filtered DataFrame is empty."""
```
Comment (indiah444) [inline streamlit_dashboard/test_dashboard.py:114]: Good situation to test

File: streamlit_dashboard/test_dashboard.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+            FROM users
+            WHERE email_address = %s AND password = %s
+        );
+        """,
+        ('test@example.com', 'password123')
+    )
+
+
+def test_login_failed_incorrect_username(mock_db_connection):
+    """Tests login failure if incorrect password is entered."""
```
Comment (indiah444) [inline streamlit_dashboard/test_dashboard.py:149]: Good testing of mistakes users could make on the dashboard

**Review by indiah444 — DISMISSED** [review]
Love all of these tests, thank you for writing them!

**Review by indiah444 — APPROVED** [review]
Good work :)
